File: apps/chat-service/websocket.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
	CheckOrigin: func(r *http.Request) bool {
		return true
	},
}

type Event struct {
	Type string      `json:"type"`
	Data interface{} `json:"data"`
}

type Client struct {
	hub      *WebSocketHub
	conn     *websocket.Conn
	send     chan []byte
	channels map[string]bool
}

type WebSocketHub struct {
	clients    map[*Client]bool
	broadcast  chan []byte
	register   chan *Client
	unregister chan *Client
	mu         sync.RWMutex
	running    bool
}

func NewWebSocketHub() *WebSocketHub {
	return &WebSocketHub{
		clients:    make(map[*Client]bool),
		broadcast:  make(chan []byte, 256),
		register:   make(chan *Client),
		unregister: make(chan *Client),
	}
}

func (h *WebSocketHub) Run() {
	h.mu.Lock()
	h.running = true
	h.mu.Unlock()

	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case client := <-h.register:
			h.mu.Lock()
			h.clients[client] = true
			h.mu.Unlock()
			log.Printf("WebSocket client connected. Total: %d", len(h.clients))

		case client := <-h.unregister:
			h.mu.Lock()
			if _, ok := h.clients[client]; ok {
				delete(h.clients, client)
				close(client.send)
			}
			h.mu.Unlock()
			log.Printf("WebSocket client disconnected. Total: %d", len(h.clients))

		case message := <-h.broadcast:
			h.mu.RLock()
			for client := range h.clients {
				select {
				case client.send <- message:
				default:
					close(client.send)
					delete(h.clients, client)
				}
			}
			h.mu.RUnlock()

		case <-ticker.C:
			h.mu.RLock()
			for client := range h.clients {
				if err := client.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
					client.conn.Close()
					delete(h.clients, client)
				}
			}
			h.mu.RUnlock()
		}
	}
}

func (h *WebSocketHub) Stop() {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.running = false
	for client := range h.clients {
		client.conn.Close()
		close(client.send)
	}
	h.clients = make(map[*Client]bool)
}

func (h *WebSocketHub) Broadcast(event Event) {
	data, err := json.Marshal(event)
	if err != nil {
		log.Printf("Failed to marshal event: %v", err)
		return
	}
	h.broadcast <- data
}

func (h *WebSocketHub) ClientCount() int {
	h.mu.RLock()
	defer h.mu.RUnlock()
	return len(h.clients)
}

func (c *Client) readPump() {
	defer func() {
		c.hub.unregister <- c
		c.conn.Close()
	}()

	c.conn.SetReadLimit(512)
	c.conn.SetReadDeadline(time.Now().Add(60 * time.Second))
	c.conn.SetPongHandler(func(string) error {
		c.conn.SetReadDeadline(time.Now().Add(60 * time.Second))
		return nil
	})

	for {
		_, message, err := c.conn.ReadMessage()
		if err != nil {
			if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
				log.Printf("WebSocket read error: %v", err)
			}
			break
		}

		var msg struct {
			Type     string   `json:"type"`
			Channel  string   `json:"channel,omitempty"`
			Channels []string `json:"channels,omitempty"`
		}
		if err := json.Unmarshal(message, &msg); err != nil {
			continue
		}

		switch msg.Type {
		case "subscribe":
			if msg.Channel != "" {
				c.channels[msg.Channel] = true
			}
		case "unsubscribe":
			if msg.Channel != "" {
				delete(c.channels, msg.Channel)
			}
		case "subscribe:channels":
			for _, ch := range msg.Channels {
				c.channels[ch] = true
			}
		}
	}
}

func (c *Client) writePump() {
	ticker := time.NewTicker(30 * time.Second)
	defer func() {
		ticker.Stop()
		c.conn.Close()
	}()

	for {
		select {
		case message, ok := <-c.send:
			c.conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if !ok {
				c.conn.WriteMessage(websocket.CloseMessage, []byte{})
				return
			}

			w, err := c.conn.NextWriter(websocket.TextMessage)
			if err != nil {
				return
			}
			w.Write(message)

			n := len(c.send)
			for i := 0; i < n; i++ {
				w.Write([]byte{'\n'})
				w.Write(<-c.send)
			}

			if err := w.Close(); err != nil {
				return
			}

		case <-ticker.C:
			c.conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
			if err := c.conn.WriteMessage(websocket.PingMessage, nil); err != nil {
				return
			}
		}
	}
}

func (h *ServiceHandler) handleWebSocket(w http.ResponseWriter, r *http.Request) {
	if !h.service.config.HasStaticKey {
		writeError(w, http.StatusForbidden, "static key not available")
		return
	}

	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		log.Printf("WebSocket upgrade error: %v", err)
		return
	}

	client := &Client{
		hub:      h.service.wsHub,
		conn:     conn,
		send:     make(chan []byte, 256),
		channels: make(map[string]bool),
	}

	h.service.wsHub.register <- client

	go client.writePump()
	go client.readPump()
}

File: apps/chat-service/src/websocket.ts
import { WebSocketServer, WebSocket } from 'ws';
import { Server } from 'http';

type EventHandler = (data: any) => void;

interface Client {
  ws: WebSocket;
  subscriptions: Set<string>;
}

export class WebSocketHub {
  private clients: Set<Client> = new Set();
  private wss: WebSocketServer;

  constructor(server: Server) {
    this.wss = new WebSocketServer({ server });
    this.wss.on('connection', (ws) => {
      const client: Client = { ws, subscriptions: new Set() };
      this.clients.add(client);

      ws.on('message', (data) => {
        try {
          const msg = JSON.parse(data.toString());
          if (msg.type === 'subscribe' && Array.isArray(msg.events)) {
            for (const event of msg.events) {
              client.subscriptions.add(event);
            }
          } else if (msg.type === 'unsubscribe' && Array.isArray(msg.events)) {
            for (const event of msg.events) {
              client.subscriptions.delete(event);
            }
          }
        } catch {
          ws.send(JSON.stringify({ error: 'Invalid message' }));
        }
      });

      ws.on('close', () => {
        this.clients.delete(client);
      });

      ws.send(JSON.stringify({ type: 'connected' }));
    });
  }

  broadcast(event: string, data: any): void {
    const message = JSON.stringify({ event, data, timestamp: Date.now() });
    
    for (const client of this.clients) {
      if (client.subscriptions.has(event) || client.subscriptions.has('*')) {
        if (client.ws.readyState === WebSocket.OPEN) {
          client.ws.send(message);
        }
      }
    }
  }

  broadcastAll(data: any): void {
    const message = JSON.stringify({ ...data, timestamp: Date.now() });
    
    for (const client of this.clients) {
      if (client.ws.readyState === WebSocket.OPEN) {
        client.ws.send(message);
      }
    }
  }

  getClientCount(): number {
    return this.clients.size;
  }

  close(): void {
    this.wss.close();
  }
}
